crates/wasi/src/filesystem.rs

```diff
@@ -236,45 +252,175 @@ impl Dir {
+                    Err(e) => ReadState::Error(e),
+                };
+            }
+            _ => {}
+        }
+    }
+
+    fn read_available_data(&mut self, size: usize) -> Bytes {
+        let ReadState::DataAvailable(b) = &mut self.state else {
+            unreachable!();
+        };
+        debug_assert!(b.len() > 0);
+
+        if size == 0 {
+            return Bytes::new();
+        }
+
+        let min_len = b.len().min(size);
+        let chunk = b.split_to(min_len);
+        if b.len() == 0 {
+            self.state = ReadState::Idle;
+        }
+        self.position += min_len as u64;
+        chunk
+    }
+}
+#[async_trait::async_trait]
+impl HostInputStream for FileInputStream {
+    fn read(&mut self, size: usize) -> StreamResult<Bytes> {
+        match self.state {
+            ReadState::Idle => {
+                if size == 0 {
+                    return Ok(Bytes::new());
+                }
+
+                let p = self.position;
+                self.state = ReadState::Waiting(
+                    self.file
+                        .spawn_blocking(move |f| Self::blocking_read(f, p, size)),
+                );
+                Ok(Bytes::new())
+            }
+            ReadState::DataAvailable(_) => Ok(self.read_available_data(size)),
+            ReadState::Waiting(_) => Ok(Bytes::new()),
+            ReadState::Error(_) => match mem::replace(&mut self.state, ReadState::Closed) {
+                ReadState::Error(e) => Err(StreamError::LastOperationFailed(e.into())),
+                _ => unreachable!(),
+            },
+            ReadState::Closed => Err(StreamError::Closed),
+        }
+    }
+    /// Specialized blocking_* variant to bypass tokio's task spawning & joining
+    /// overhead on synchronous file I/O.
+    async fn blocking_read(&mut self, size: usize) -> StreamResult<Bytes> {
+        self.wait_ready().await;
+
+        match &mut self.state {
+            ReadState::Idle => {
+                if size == 0 {
+                    return Ok(Bytes::new());
+                }
+
+                let p = self.position;
+                match self
+                    .file
+                    .run_blocking(move |f| Self::blocking_read(f, p, size))
+                    .await
+                {
+                    Ok(b) if b.len() == 0 => {
+                        self.state = ReadState::Closed;
+                        Err(StreamError::Closed)
+                    }
+                    Ok(b) => {
+                        self.position += b.len() as u64;
+                        Ok(b)
+                    }
+                    Err(e) => {
+                        self.state = ReadState::Closed;
+                        Err(StreamError::LastOperationFailed(e.into()))
+                    }
+                }
+            }
+            ReadState::DataAvailable(_) => Ok(self.read_available_data(size)),
+            ReadState::Waiting(_) => Ok(Bytes::new()),
```

Comment:
Should this perhaps `await` the result of the read here inline?

Reply:
Hmm, this should've actually been `unreachable!()` as this state is awaited higher up in the function.